providers/implementations/keymgmt/template_kmgmt.c

```diff
@@ -0,0 +1,437 @@
+            /* validate whether the private keys match */
+        }
+        ok = ok && key_checked;
+    }
+    debug_print("match result %d\n", ok);
+    return ok;
+}
+
+static int key_to_params(void *key, OSSL_PARAM_BLD *tmpl,
+                         OSSL_PARAM params[], int include_private)
+{
+    if (key == NULL)
+        return 0;
+
+    /* add public and/or private key parts to templ as possible */
+
+    return 1;
+}
+
+static int template_export(void *key, int selection, OSSL_CALLBACK *param_cb,
+                           void *cbarg)
+{
+    OSSL_PARAM_BLD *tmpl;
+    OSSL_PARAM *params = NULL;
+    int ret = 0;
+
+    debug_print("export %p\n", key);
+    if (!ossl_prov_is_running() || key == NULL)
+        return 0;
+
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
+        return 0;
+
+    tmpl = OSSL_PARAM_BLD_new();
+    if (tmpl == NULL)
+        return 0;
+
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0) {
+        int include_private = ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0);
+
+        if (!key_to_params(key, tmpl, NULL, include_private))
+            goto err;
+    }
+
+    params = OSSL_PARAM_BLD_to_param(tmpl);
+    if (params == NULL)
+        goto err;
+
+    ret = param_cb(params, cbarg);
+    OSSL_PARAM_free(params);
+err:
+    OSSL_PARAM_BLD_free(tmpl);
+    debug_print("export result %d\n", ret);
+    return ret;
+}
+
+static int ossl_template_key_fromdata(void *key,
+                                      const OSSL_PARAM params[],
+                                      int include_private)
+{
+    const OSSL_PARAM *param_priv_key = NULL, *param_pub_key;
+
+    if (key == NULL)
+        return 0;
+    if (ossl_param_is_empty(params))
+        return 0;
+
+    /* validate integrity of key (algorithm type specific) */
+
+    param_pub_key = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PUB_KEY);
+    if (include_private)
+        param_priv_key =
+            OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PRIV_KEY);
+
+    if (param_pub_key == NULL && param_priv_key == NULL)
+        return 0;
+
+    if (param_priv_key != NULL) {
+        /* retrieve private key and check integrity */
+    }
+
+    if (param_pub_key != NULL) {
+        /* retrieve public key and check integrity */
+    }
+
+    return 1;
+}
+
+static int template_import(void *key, int selection, const OSSL_PARAM params[])
+{
+    int ok = 1;
+    int include_private;
+
+    debug_print("import %p\n", key);
+    if (!ossl_prov_is_running() || key == NULL)
+        return 0;
+
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
+        return 0;
+
+    include_private = selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY ? 1 : 0;
+    ok = ok && ossl_template_key_fromdata(key, params, include_private);
+
+    debug_print("import result %d\n", ok);
+    return ok;
+}
+
+#define TEMPLATE_KEY_TYPES()                                 \
+    OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_PUB_KEY, NULL, 0),   \
```

Comment:
nit: Seems like the trailing `\` should line up

Reply:
done